docs/rfcs/0004-agent-flow.md

```diff
@@ -0,0 +1,54 @@
+# Agent Flow - Agent Utilizing Components 
+
+* Date: 2022-03-30
+* Author: Matt Durham (@mattdurham)
+* PR: [grafana/agent#1538](https://github.com/grafana/agent/pull/1538)
+* Status: Draft
+
+## Overarching Problem Statement
+
+The Agents configuration and onboarding is difficult to use. Viewing the effect of configuration changes on telemetry data is difficult. Making the configuration simplier, composable and intuitive allieviates these concerns.
+
+
+## Description
+
+Agent Flow is intended to solve real world needs that the Grafana Agent team have idenfified in conversations with users and developers. 
+
+These broadly include 
+
+- Current configuration is complex and opaque
+- Lack of introspection within the agent
+    - Questions about what telemetry data are being sent
+    - Are rules applying correctly?
+    - How does filtering work?
+- Users have been requesting additional capabilites and adding new features is hard due to coupling between systems, some examples include
+    - Remote Write, Different Input Formats
+    - Different output formats
+    - Filtering ala Relabel Configs is complex and hard to figure out when they occur
+- Lack of understanding how telemetry data moves through agent
+    - Other systems use pipeline/extensions to allow users to understand how data moves through the system
+
+# 1. Introduction and Goals 
+
+This design document describes Agent Flow, which is a system for describing the execution and configuration of telemetry data, along with any secondary configuration required to configure the Grafana Agent. 
+
+Agent Flow refers to both the execution, configuration and visual configurator of data flow.
+
+The primary goals of Agent Flow are
+
+1. Allow users to easily configure the system
+2. Allow users to understand the system
+3. Allow users to inspect the system
+4. Allow developers to easily add components to the system
+
+Goals of this design document
```

Comment:
I'd love it if we could talk a little about target platforms here: we should be careful that whatever we design is performant enough to still run on embedded hardware (eg Raspberry Pi).

Maybe the goal statement should be something like

> Have 100% compatibility with existing installation targets for Grafana Agent

Reply:
(Sorry, I meant additional goal statement, not to change the entire set of goals with the above)

Reply:
Agreed! I think a composable may have a slight performance decrease but it should still run on our target platforms.